src/cluster_asm.c

```diff
@@ -119,13 +131,50 @@ char *asmTaskStateToString(int state) {
     return NULL; /* Unreachable */
 }
 
+void asmTaskReset(asmTask *task) {
+    task->state = ASM_NONE;
+    task->rdb_channel_state = ASM_NONE;
+    task->main_channel_id = -1;
+    task->main_channel_conn = NULL;
+    task->rdb_channel_conn = NULL;
+    task->dest_offset = -1;
+    task->source_offset = -1;
+    replDataBufInit(&task->sync_buffer);
+}
+
+asmTask *asmTaskCreate(void) {
+    asmTask *task = zcalloc(sizeof(*task));
+    asmTaskReset(task);
+    task->slot_ranges = NULL;
+    task->source_node = NULL;
+    task->main_channel_client = NULL;
+    task->retry_count = 0;
+    task->create_time = server.mstime;
+    task->start_time = 0;
+    task->done_time = 0;
+
+    return task;
+}
+
+/* Compare two slot range arrays, return 1 if equal, 0 otherwise */
+static int slotRangeArrayIsEqual(slotRangeArray *sra1, slotRangeArray *sra2) {
+    if (sra1->num_ranges != sra2->num_ranges) return 0;
+    for (int i = 0; i < sra1->num_ranges; i++) {
+        if (sra1->ranges[i].start != sra2->ranges[i].start ||
+            sra1->ranges[i].end != sra2->ranges[i].end) {
+            return 0;
+        }
+    }
+    return 1;
```

Comment:
I have the same function in my local :) We'll have to match slotranges to check if there is ongoing task in many places. Though, I think we'll need to sort ranges first. It can be given out of order. I had this on my local, I guess we can call it just before comparison:

```c
static int compareSlotRanges(const void *a, const void *b) {
    const slotRange *sa = a;
    const slotRange *sb = b;
    if (sa->start < sb->start) return -1;
    if (sa->start > sb->start) return 1;
    return 0;
}

void sortSlotRanges(slotRangeArray *sra) {
    qsort(sra->ranges, sra->num_ranges, sizeof(slotRange), compareSlotRanges);
}
```

Reply:
make great sense, should I apply your suggestions?